src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

```diff
@@ -9,94 +10,170 @@ internal struct FileStatus
-            {
-                // User effectively owns the file
-                readBit = Interop.Sys.Permissions.S_IRUSR;
-                writeBit = Interop.Sys.Permissions.S_IWUSR;
+                if (_fileCache.Uid == (_euid ??= Interop.Sys.GetEUid()))
+                {
+                    // User effectively owns the file
+                    readBit = Interop.Sys.Permissions.S_IRUSR;
+                    writeBit = Interop.Sys.Permissions.S_IWUSR;
+                }
+                else if (_fileCache.Gid == (_egid ??= Interop.Sys.GetEGid()))
+                {
+                    // User belongs to a group that effectively owns the file
+                    readBit = Interop.Sys.Permissions.S_IRGRP;
+                    writeBit = Interop.Sys.Permissions.S_IWGRP;
+                }
+                else
+                {
+                    // Others permissions
+                    readBit = Interop.Sys.Permissions.S_IROTH;
+                    writeBit = Interop.Sys.Permissions.S_IWOTH;
+                }
+#endif
+
+                return (_fileCache.Mode & (int)readBit)  != 0 && // has read permission
+                       (_fileCache.Mode & (int)writeBit) == 0;   // but not write permission
             }
-            else if (_fileStatus.Gid == Interop.Sys.GetEGid())
+        }
+
+        // Checks if the main path is a symbolic link
+        // Only call if Refresh has been successfully called at least once
+        private bool HasSymbolicLinkFlag
+        {
+            get
             {
-                // User belongs to a group that effectively owns the file
-                readBit = Interop.Sys.Permissions.S_IRGRP;
-                writeBit = Interop.Sys.Permissions.S_IWGRP;
+                Debug.Assert(IsFileCacheInitialized);
+                return (_fileCache.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFLNK;
             }
-            else
+        }
+
+        // Sets the cache initialization flags to -1, which means the caches are now uninitialized
+        internal void InvalidateCaches()
+        {
+            _initializedFileCache = -1;
+            _initializedSymlinkCache = -1;
+        }
+
+        internal bool IsReadOnly(ReadOnlySpan<char> path, bool continueOnError = false)
+        {
+            EnsureCachesInitialized(path, continueOnError);
+            return HasReadOnlyFlag;
+        }
+
+        internal bool IsHidden(ReadOnlySpan<char> path, ReadOnlySpan<char> fileName, bool continueOnError = false)
+        {
+            // Avoid disk hit first
+            if (IsNameHidden(fileName))
             {
-                // Others permissions
-                readBit = Interop.Sys.Permissions.S_IROTH;
-                writeBit = Interop.Sys.Permissions.S_IWOTH;
+                return true;
             }
-#endif
+            EnsureCachesInitialized(path, continueOnError);
+            return HasHiddenFlag;
+        }
```

Comment:
IIRC you mentioned that UF_HIDDEN is only effective in OSX, if so we should consider splitting this implementation and optimize Unix by avoiding the native call.